### PR TITLE
Added Drag Drop support to allow dragging objects into the browser

### PR DESCRIPTION
**Pull request by Bodekaer**
Supports files, links, text, html etc.
Currently only supported in WPF (would be easy to add WinForm support though, by just copy-pasting the WPF ChromiumBrowser code)

FYI #171

### CODE REVIEW

File: CefSharp/CefDragOperationsMask.cs [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+using System;
+
+namespace CefSharp
+{
+    /// <summary>
+    /// Drag operation mask types
+    /// </summary>
+    [Flags]
+    public enum CefDragOperationsMask
```
Comment (amaitland) [inline CefSharp/CefDragOperationsMask.cs]: There's already a `CefDragOperationsMask` enum see https://github.com/cefsharp/CefSharp/blob/cef/2171/CefSharp/DragOperationsMask.cs

Just a note most enums in `Cef` are `uint`, whilst you can cast between the two, the value for `Every` is different being `uint.MaxValue`

File: CefSharp/MouseEvent.cs [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+//
+// Use of this source code is governed by a BSD-style license that can be found in the LICENSE file.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+
+namespace CefSharp
+{
+    public class MouseEvent : IMouseEvent
```
Comment (amaitland) [inline CefSharp/MouseEvent.cs]: Do we need the `IMouseEvent` interface? Can probably make this a struct as well (There's a few other classes in the code base that probably should be structs, one day I'll get around to changing them)

File: CefSharp.Wpf/ChromiumWebBrowser.cs [hunk 3]
```diff
@@ -459,6 +467,151 @@ public ChromiumWebBrowser()
             PresentationSource.AddSourceChangedHandler(this, PresentationSourceChangedHandler);
         }
 
+        void OnDrop(object sender, DragEventArgs e)
```
Comment (amaitland) [inline CefSharp.Wpf/ChromiumWebBrowser.cs]: Can you please add the `private` keyword to these methods for consistency :+1:

File: CefSharp.Wpf/ChromiumWebBrowser.cs [hunk 4]
```diff
@@ -422,6 +424,12 @@ public ChromiumWebBrowser()
             GotKeyboardFocus += OnGotKeyboardFocus;
             LostKeyboardFocus += OnLostKeyboardFocus;
 
+            // Drag Drop events
+            DragEnter += OnDragEnter;
```
Comment (amaitland) [inline CefSharp.Wpf/ChromiumWebBrowser.cs]: I see these event handlers being `hooked`, are they `unhooked` in dispose?

File: CefSharp.Core/Internals/CefDragDataWrapper.h [hunk 5]
```diff
@@ -41,6 +41,24 @@ namespace CefSharp
...
+            // Create a new CefDragData object.
+            ///
+            /*--cef()--*/
+            static CefDragDataWrapper^ Create(){
+                CefRefPtr<CefDragData> cefDragData = CefDragData::Create();
+                return gcnew CefDragDataWrapper(cefDragData);
+            }
+
+
+            virtual property CefRefPtr<CefDragData>* _internalDragData
```
Comment (amaitland) [inline CefSharp.Core/Internals/CefDragDataWrapper.h]: Not really a fan of naming a property starting with a `_`, calling it `InternalDragData` should be fine.

**Comment by amaitland** [comment]
@Bodekaer Looks pretty solid overall :+1: I've commented inline where appropriate

In general there's some cleanup that still needs to happen
- Opening braces (`{`) on new lines
- Cleanup whitespace

> Currently only supported in WPF (would be easy to add WinForm support though, by just copy-pasting the WPF ChromiumBrowser code)

FYI `WinForms` version already has drag and drop built in, I'm pretty sure those methods are only required for `OSR`

File: CefSharp/MouseEvent.cs (hunk 2 above)
Comment (amaitland) [inline CefSharp/MouseEvent.cs]: Don't worry about the `struct` idea, we'll just keep things consistent for now.

Comment (Bodekaer) [inline CefSharp/MouseEvent.cs]: Thanks for all the great feedback Alex!
I'll run through it asap and update the PR.

On 1/8/2015 10:39:25 AM, Alex Maitland <notifications@github.com> wrote:
In CefSharp/MouseEvent.cs [https://github.com/cefsharp/CefSharp/pull/732#discussion-diff-22642543]:
> @@ -0,0 +1,17 @@ > +﻿// Copyright © 2010-2014 The CefSharp Authors. All rights reserved. > +// > +// Use of this source code is governed by a BSD-style license that can be found in the LICENSE file. > + > +using System; > +using System.Collections.Generic; > +using System.IO; > + > +namespace CefSharp > +{ > + public class MouseEvent : IMouseEvent
Don't worry about the struct idea, we'll just keep things consistent for now.

**Comment by amaitland** [comment]
@Bodekaer When you have a sec can you retarget this against `master`? (I think you should just be able to edit the target branch above). I've merged `cef/2171` into `master` :smile:

**Comment by Bodekaer** [comment]
Yep definitely. I just created a new `PR` against `master` here #756.
I cannot edit the comparison of this PR, so we can just close it.

Just FYI, I've been hyper busy lately, so I haven't yet cleaned up the code based on your feedback.

**Comment by amaitland** [comment]
@Bodekaer Thanks!

**Pull request closed without merging**